gpt_engineer/core/default/steps.py

```diff
@@ -0,0 +1,257 @@
+import inspect
+import os
+import re
+from termcolor import colored
+import subprocess
+
+#TODO: THIS NEEDS A BETTER SOLUTION
+PREPROMPTS_PATH = os.path.join("gpt_engineer", "preprompts")
+
+
+
+def curr_fn() -> str:
+    """
+    Retrieves the name of the calling function.
+
+    This function uses Python's inspection capabilities to dynamically fetch the
+    name of the function that called `curr_fn()`. This approach ensures that the
+    function's name isn't hardcoded, making it more resilient to refactoring and
+    changes to function names.
+
+    Returns:
+    - str: The name of the function that called `curr_fn()`.
+    """
+    return inspect.stack()[1].function
+def setup_sys_prompt(db: FileRepository) -> str:
+    """
+    Constructs a system prompt for the AI based on predefined instructions and philosophies.
+
+    This function is responsible for setting up the system prompts for the AI, instructing
+    it on how to generate code and the coding philosophy to adhere to. The constructed prompt
+    consists of the "roadmap", "generate" (with dynamic format replacements), and the coding
+    "philosophy" taken from the given DBs object.
+
+    Parameters:
+    - dbs (DBs): The database object containing pre-defined prompts and instructions.
+
+    Returns:
+    - str: The constructed system prompt for the AI.
+    """
+    return (
+        db["roadmap"]
+        + db["generate"].replace("FILE_FORMAT", db["file_format"])
+        + "\nUseful to know:\n"
+        + db["philosophy"]
+    )
+
+
+def gen_code(workspace_path: str, ai: AI, prompt: str, memory: FileRepository) -> Code:
+    """
+    Executes the AI model using the default system prompts and saves the full output to memory and program to disk.
+
+    This function prepares the system prompt using the provided database configurations
+    and then invokes the AI model with this system prompt and the main input prompt.
+    Once the AI generates the output, this function saves it to the specified workspace.
+    The AI's execution is tracked using the name of the current function for contextual reference.
+
+    Parameters:
+    - ai (AI): An instance of the AI model.
+    - dbs (DBs): An instance containing the database configurations, including system and
+      input prompts, and file formatting preferences.
+
+    Returns:
+    - List[Message]: A list of message objects encapsulating the AI's generated output.
+
+    Note:
+    The function assumes the `ai.start` method and the `to_files` utility are correctly
+    set up and functional. Ensure these prerequisites are in place before invoking `simple_gen`.
+    """
+    db = FileRepository(PREPROMPTS_PATH)
+    messages = ai.start(setup_sys_prompt(db), prompt, step_name=curr_fn())
+    chat = messages[-1].content.strip()
+    memory[CODE_GEN_LOG_FILE] = chat
+    files = parse_chat(chat)
+    workspace = FileRepository(workspace_path)
+    for file_name, file_content in files:
+        workspace[file_name] = file_content
```

Comment:
why do we write to disk here?

I thought a big idea behind this refactor was that we abstract the IO and only modify a Code object in memory.

Reply:
![image](https://github.com/AntonOsika/gpt-engineer/assets/4467025/82634ff3-9977-4cf5-9ae3-bf53cd08044b)